### ISSUE TITLE
clairctl can't be used for Amazon ECR Registries

### ISSUE THREAD

**Issue by lcgkm** [jgsqware/clairctl#42]
How to config the username/password to Log in to a Docker registry?

**Comment by jgsqware** [comment]
Currently, the authentification is donne by docker login command.
You logged in first via the docker command, than your run clairctl
command.

How the Amazon ECR authentication is working?

**Comment by lcgkm** [comment]
OK. Got it. Thank you!

>How the Amazon ECR authentication is working?
We can login AWS and got a docker login username and password
```
%aws ecr get-login
docker login -u AWS -p password -e none https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com
```
http://docs.aws.amazon.com/AmazonECR/latest/userguide/Registries.html

**Comment by lcgkm** [comment]
@jgsqware 
Not worked. I have login via docker command and run clairctl command at the same terminal
```
% docker pull xxxxxx.dkr.ecr.us-east-1.amazonaws.com/ops/heapster:1.2.0-1                  
1.2.0-1: Pulling from ops/heapster
444873b88xxx: Downloading [===========>                                       ] 590.3 kB/2.567 MB
ae3b6e76axxx: Pulling fs layer 
1ce8e1087xxx: Downloading [=>                                                 ] 468.5 kB/14.04 MB
afd3869c5xxx: Waiting
......
% ./clairctl pull xxxxxx.dkr.ecr.us-east-1.amazonaws.com/ops/heapster:1.2.0-1 --log-level Debug
2017-05-19 14:23:42.654075 D | config: Using config file: /Users/xxxxxx/.go/src/github.com/jgsqware/clairctl/clairctl.yml
2017-05-19 14:23:42.654290 D | dockerdist: Downloading manifest for xxxxxx.dkr.ecr.us-east-1.amazonaws.com/ops/heapster:1.2.0-1
2017-05-19 14:23:42.654400 D | dockerdist: Retrieving repository client
2017-05-19 14:23:44.338139 D | dockerdist: Auth: err: Get https://xxxxxx.dkr.ecr.us-east-1.amazonaws.com/v2/: no basic auth credentials
client quit unexpectedly
2017-05-19 14:23:44.338180 C | cmd: retrieving manifest for "xxxxxx.dkr.ecr.us-east-1.amazonaws.com/ops/heapster:1.2.0-1": Get https://741589461034.dkr.ecr.us-east-1.amazonaws.com/v2/: no basic auth credentials
```

**Comment by jgsqware** [comment]
Ok cool I learned something. I should add a log for the login when running
clairctl

Is it working for you now?

**Comment by lcgkm** [comment]
Not worked.
```
2017-05-19 14:23:44.338139 D | dockerdist: Auth: err: Get https://xxxxxx.dkr.ecr.us-east-1.amazonaws.com/v2/: no basic auth credentials
client quit unexpectedly
2017-05-19 14:23:44.338180 C | cmd: retrieving manifest for "xxxxxx.dkr.ecr.us-east-1.amazonaws.com/ops/heapster:1.2.0-1": Get https://741589461034.dkr.ecr.us-east-1.amazonaws.com/v2/: no basic auth credentials

```

**Comment by jgsqware** [comment]
Pull work but not push, I need to implement the go aws client to retrieve the downloadable url.

**Comment by goffinf** [comment]
I definitely have a requirement to be able to integrate clairctl with ECR since this is our primary on-prem private repository.

As far as login is concerned, we configure the ECR Credentials Helper on all instances. As you may know this seemlessly handles login to ECR and returns an authentication token (which is then cached and is valid for 12 hours - after which, if a new request is made, the process is repeated and a new token retrieved). Of course there still remains the question of the policy(ies) that are applied to ECR repos vs the requesting user privileges (we use an IAM instance profile here which assumes a specific role).

Let me know if there are further details that might be helpful, or if you want me to test an specific use cases.

Regards

Fraser.

**Comment by jgsqware** [comment]
Thanks for this. Can it be ok for if clairctl use aws CLI under the hood?
That's means you have to have aws CLI installed on your machine

**Comment by goffinf** [comment]
Yep you should assume the AWS CLI is a pre-req. We install it on all EC2 instances as part of the standard build, and when we containerize apps that make use of the CLI we also install it into the image and then find cunningways of mapping the credentials into the container at launch.

**Comment by sbutt** [comment]
Hi, this would be really useful for us. Is there any progress with this? Thanks, Steve

**Comment by ggulati2** [comment]
Any progress here?

**Comment by goffinf** [comment]
I haven’t seen any update that would suggest that clairctl supports ECR auth.

We still would very much like this but all of our recent attempts have failed.

It’s really straight forwards to configure any container, including clairctl to engage the ECR credentials helper I mentioned above. When we do that you can docker exec into clairctl and issue a docker pull ecr-namespace/image:tag and it will work successfully. But trying the same thing with clairctl pull ecr-namespace/image:tag always fails with ‘No basic auth ... blah’. My knowledge of Go is pretty limited but, looking at the source it would appear that there is no current support.

For now we have abandoned trying to get that to work in favour of doing an explicit docker pull before calling clairctl analyze. After the pull we either address the image directly using —local or we push to a local registry container and the call clairctl analyze using the local registry namespace. Since this is all part of a Jenkins pipeline it still works pretty well, but for sure I would prefer it if I could just go direct to ECR images with clairctl pull/analyze/whatever.

Fraser.

**Comment by PeteGoo** [comment]
I've added ECR integration in our fork using the AWS SDK. Is this something you would entertain being added here?
https://github.com/jgsqware/clairctl/compare/master...PeteGoo:ECR
